### ISSUE TITLE
Phpstornm (Ide) Autocomplete is broken for Custom QueryBuilder (@return self vs static)

### ISSUE THREAD

**Issue by c33s** [doctrine/orm#7868]
**TLDR:** can we change all `@return self` to `@return static` in `lib/Doctrine/ORM/QueryBuilder.php` to fix typehinting in custom querybuilder?

### Feature Request / Kind of Bug Report

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | yes
| BC Break    | don't know

if i create a custom `PersonQueryBuilder` class which extends the doctrine `QueryBuilder` the typehinting is broken. 

`Return value is expected to be 'PersonQueryBuilder', '\Doctrine\ORM\QueryBuilder' returned`
(violett marks error lines)

![doctrine-static-vs-self](https://user-images.githubusercontent.com/649209/66832072-64931780-ef59-11e9-9e2e-4756579e3627.png)

#### Summary

the reason is because of the `@return self` annotation in the docblock of the `QueryBilder` class. if i change it to `@return static` the typehinting is correct again.

`QueryBuilder`
```
...
    /**
     * Sets the maximum number of results to retrieve (the "limit").
     *
     * @param integer|null $maxResults The maximum number of results to retrieve.
     *
     * @return self       <----- changing this to `@return static` makes it work
     */
    public function setMaxResults($maxResults)
    {
        $this->_maxResults = $maxResults;

        return $this;
    }
...
```

#### Current workaround
add `@method` annotations to the custom querybuilder class for all methods you need direct or indirect (for example `orderBy` also calls `add`)

```
<?php


namespace App\Doctrine\ORM\QueryBuilder;

use Doctrine\ORM\QueryBuilder;


/**
 * Class PersonQueryBuilder
 * @package App\Doctrine\ORM\QueryBuilder
 *
 * @method PersonQueryBuilder orderBy($sort, $order = null)
 * @method PersonQueryBuilder add($dqlPartName, $dqlPart, $append = false)
 * @method PersonQueryBuilder addOrderBy($sort, $order = null)
 * @method PersonQueryBuilder innerJoin($join, $alias, $conditionType = null, $condition = null, $indexBy = null)
 * @method PersonQueryBuilder addSelect($select = null)
 * @method PersonQueryBuilder select($select = null)
 * @method PersonQueryBuilder from($from, $alias, $indexBy = null)
 * @method PersonQueryBuilder andWhere($condition)
 * @method PersonQueryBuilder setParameter($key, $value, $type = null)
 */
class PersonQueryBuilder extends QueryBuilder
{
    public function addSortByRank(): self
    {
        return $this
            ->orderBy('person_group.rank')
            ->addOrderBy('person.firstName', 'ASC')
            ->addOrderBy('person.lastName', 'ASC')
            ->addOrderBy('person.name', 'ASC')
        ;
    }

    public function addFilterByVisible(): self
    {
        return $this
            ->andWhere('person_group.visible = :group_is_visible')
            ->andWhere('person.visible = :person_is_visible')
            ->setParameter('group_is_visible', true)
            ->setParameter('person_is_visible', true)
        ;
    }
}
```

**Comment by Ocramius** [comment]
This is the correct behavior: the query builder is not designed for inheritance in first place.

**Comment by c33s** [comment]
@Ocramius maybe you want to tell us what the best practice is to add custom commands to the querybuilder? https://stackoverflow.com/questions/58357664/what-is-the-best-practice-to-add-custom-methods-to-the-querybuilder-based-upon-e

does it make sense to not allow/design the querybuilder to be extended? if i can define a custom querybuilder class in my repo class, it makes more that sense to just be able to add one or to methods in an extended querybuilder instead of copying the whole class and hope it does not break with the next update.

**Comment by Ocramius** [comment]
In theory, `QueryBuilder` should be `final`: in practice, we can no longer do that due to BC boundaries.

The correct approach to custom query builders is to indeed ship your own, since the builder is responsible for its internal consistency, and that consistency changes with the DSL you want to support.
